File: src/primitives/nodes.jl
"""
    MARKERS

The stock marker glyphs a node family may be drawn with. These are the names the package owns and
validates. They are not the whole set a [`Nodes`](@ref) family may name: `marker` also takes an
image of your own, and the name of a sprite a browser module registered.
"""
const MARKERS = (:disc, :square, :diamond, :triangle, :triangle_down, :triangle_right,
                 :triangle_left, :pentagon, :hexagon, :star, :cross, :x)

"""
    marker_image(path) -> String

Read an image file and return the `data:` URI a [`Nodes`](@ref) family takes for `marker`. PNG and
SVG are the two formats every browser draws; the MIME type comes from the extension.

Give the image a square canvas. A billboard reads `size` for its width and its height together, so a
wide image is drawn squeezed.

Use a `data:` URI and not a link to an image on a server. The viewer runs under a policy that admits
the first and refuses the second, and a refused image draws nothing at all.

```julia
Nodes(:sat; position = pos, marker = marker_image("assets/satellite.png"), size = 20)
```
"""
function marker_image(path)
    ext = lowercase(splitext(String(path))[2])
    mime = ext == ".svg" ? "image/svg+xml" :
           ext == ".png" ? "image/png" :
           ext in (".jpg", ".jpeg") ? "image/jpeg" :
           ext == ".webp" ? "image/webp" :
           throw(ArgumentError("marker_image draws .png, .svg, .jpg or .webp (got $(repr(ext)))"))
    return "data:$mime;base64," * base64encode(read(path))
end

# Where a label sits against the point the offset lands on: across the text, then down it.
const LABEL_ACROSS = (:left, :center, :right)
const LABEL_DOWN = (:top, :center, :bottom, :baseline)

# A range of camera distances, in the two shapes the viewer reads one in: `(near_m, far_m)` for a
# visibility range, and four values for a ramp that runs between the two distances. `form` names the
# shape in the error, since the two four-value fields ramp different things. The near distance must
# stay below the far one. Cesium refuses a display condition whose range does not grow, and a ramp
# over such a range has nothing to interpolate.
near_far(what, form, ::Nothing, n) = nothing
function near_far(what, form, v, n)
    v isa Union{Tuple,AbstractVector} && length(v) == n && all(x -> x isa Real, v) ||
        throw(ArgumentError("$what is $form (got $(repr(v)))"))
    far = n == 2 ? v[2] : v[3]
    v[1] < far || throw(ArgumentError(
        "$what needs its near metres below its far ones (got $(v[1]) and $far)"))
    return Tuple(Float64(x) for x in v)
end

"""
    Label(texts; align = (:left, :bottom), offset = (0, -14), font = "13px sans-serif",
          color = nothing, background = nothing, scale_by_distance = nothing,
          fade_by_distance = nothing, show_between = nothing)

The texts a [`Nodes`](@ref) family writes beside its entities, and the look it writes them in. Pass
one as the `label` keyword of `Nodes`. `texts` holds one string per entity. Every other field covers
the whole family. A label is a style and never a position. A plain `Vector{String}` in that keyword
is this style with every default kept.

Each label follows the node it belongs to. `offset` moves it from there in pixels, `x` to the right
and `y` down, and `align` says which part of the text lands on that point. `(:left, :bottom)` puts
the bottom left corner of the text there, so the text hangs up and to the right of the node.
`(:center, :bottom)` centres the text over the node. `font` is a CSS font string.

`color` fills the glyphs over the black outline every label keeps, and takes the colour forms
`Nodes` takes for `color`. `background` draws a box behind the text in the colour you give it.
Without it the text stands on the scene.

Three fields read the camera distance, in metres. `scale_by_distance` is
`(near_m, near_scale, far_m, far_scale)`, `fade_by_distance` is
`(near_m, near_alpha, far_m, far_alpha)`, and `show_between` is `(near_m, far_m)`, the distances the
text is drawn between. All three need the near metres below the far ones.

```julia
Nodes(:city; position = pos, label = Label(names; align = (:center, :bottom), offset = (0, -18)))
Nodes(:gw; position = gw, label = Label(names; color = "#ffd166", show_between = (0, 5.0e6)))
```
"""
struct Label
    text::Vector{String}
    align::Tuple{Symbol,Symbol}
    # Pixels from the node, `x` to the right and `y` down, which is the browser's screen axis.
    offset::NTuple{2,Float64}
    font::String
    color::Union{Nothing,Array{UInt8}}
    background::Union{Nothing,Array{UInt8}}
    scale_by_distance::Union{Nothing,NTuple{4,Float64}}
    fade_by_distance::Union{Nothing,NTuple{4,Float64}}
    show_between::Union{Nothing,NTuple{2,Float64}}
    # An INNER constructor, for the reason `Nodes` has one: an exact-typed call would otherwise
    # reach the auto-generated one and put an unchecked style on the wire.
    function Label(text, align, offset, font, color, background, scale_by_distance,
                   fade_by_distance, show_between)
        align isa Union{Tuple,AbstractVector} && length(align) == 2 &&
            Symbol(align[1]) in LABEL_ACROSS && Symbol(align[2]) in LABEL_DOWN ||
            throw(ArgumentError("Label.align is one of $LABEL_ACROSS across the text and one of " *
                                "$LABEL_DOWN down it (got $(repr(align)))"))
        offset isa Union{Tuple,AbstractVector} && length(offset) == 2 &&
            all(x -> x isa Real, offset) ||
            throw(ArgumentError("Label.offset is (x_px, y_px) from the node, y downward " *
                                "(got $(repr(offset)))"))
        return new(collect(String, text), (Symbol(align[1]), Symbol(align[2])),
                   NTuple{2,Float64}(offset), String(font), to_colors(color), to_colors(background),
                   near_far("Label.scale_by_distance", "(near_m, near_scale, far_m, far_scale)",
                            scale_by_distance, 4),
                   near_far("Label.fade_by_distance", "(near_m, near_alpha, far_m, far_alpha)",
                            fade_by_distance, 4),
                   near_far("Label.show_between", "(near_m, far_m)", show_between, 2))
    end
end

Label(texts; align = (:left, :bottom), offset = (0, -14), font = "13px sans-serif",
      color = nothing, background = nothing, scale_by_distance = nothing,
      fade_by_distance = nothing, show_between = nothing) =
    Label(texts, align, offset, font, color, background, scale_by_distance, fade_by_distance,
          show_between)

# The strings a `label` knob carries, in whichever of its two forms it arrives.
label_texts(::Nothing) = nothing
label_texts(l::Label) = l.text
label_texts(v) = collect(String, v)

"""
    Nodes(kind; position, color=nothing, size=nothing, marker=:disc, label=nothing,
          show=nothing, scale_by_distance=nothing)

A family of points or stock sprites. `position` is ECEF metres, `3 × N` for a family that stands
still through the window or `3 × N × count` for one whose positions are interpolated across it.

`color`, `size` and `show` follow the array convention and switch at the keyframe crossing:
a lone value covers the family, one value per entity varies across it, and a trailing keyframe
dimension varies it over time. `marker` names what each entity is drawn with: one of `$(MARKERS)` — a
white glyph the per-entity colour tints — an image of your own from [`marker_image`](@ref), an
`assets/<mount>/<file>` path the server serves, or the owner-namespaced name of a sprite a browser
module registered (`"orbits.pulse"`). The same tint multiplies a supplied image, which the default
white leaves as you drew it. `label` is one string per entity, or a [`Label`](@ref) that
carries those strings and says how they are drawn. `scale_by_distance` is
`(near_m, near_scale, far_m, far_scale)`, so markers stay legible close up and shrink when the whole
scene is in view.

**`show` is a per-entity visibility mask** — zero hides an entity, any other value draws it, and an
absent `show` draws the whole family. A `BitVector` of a predicate is the shape it is for. It is the
only knob that is not about appearance, and it exists because a family's membership is fixed for a
window's life: a keyframe changes what the entities look like, never how many there are. Masking is
how the drawn set varies within one window, and it costs one batch-table write — the same class as a
recolour, with no rebuild.

Two consequences of masking rather than dropping. A masked entity keeps its index, so an anchor or an
[`Edges`](@ref) pair naming it stays valid. And it is **not pickable** while hidden: nothing is drawn
for it, so nothing under the pointer reports it — a mask is not a dimming.

```julia
Nodes(:sat; position = pos, size = 12, color = rgba(CMAP, throughput; range = (0, 12)))
Nodes(:gw; position = gw_ecef, marker = :star, color = (200, 120, 255, 255))
```
"""
struct Nodes
    # A family name the scene author invents, so a `String` on both sides of the wire (ADR-0029):
    # a pointer event reports it back as `ev.entity.kind`.
    kind::String
    position::Array{Float32}
    color::KnobValue
    size::KnobValue
    # Per-entity visibility: zero hides, anything else draws, `nothing` draws the family.
    show::KnobValue
    # A stock glyph name, an `assets/<mount>/<file>` path, a `data:` URI, or the name of a
    # sprite a peer module registered in the browser.
    marker::String
    # A plain string array is the default style. A `Label` says what the strings look like.
    label::Union{Nothing,Vector{String},Label}
    scale_by_distance::Union{Nothing,NTuple{4,Float64}}
    # An INNER constructor so the shape checks run for every call form: an exact-typed call would
    # otherwise reach the auto-generated one and put a malformed family on the wire.
    function Nodes(kind, position, color, size, show, marker, label, scale_by_distance)
        pos = convert(Array{Float32}, position)
        ndims(pos) in (2, 3) && Base.size(pos, 1) == 3 ||
            throw(ArgumentError("$kind.position is 3 × N or 3 × N × keyframes (got $(Base.size(pos)))"))
        n = Base.size(pos, 2)
        mk = to_source(marker, "$kind.marker", MARKERS)
        texts = label_texts(label)
        texts === nothing || length(texts) == n ||
            throw(ArgumentError("$kind has $(length(texts)) labels for $n entities"))
        c, s, v = to_colors(color), to_scalars(size), to_codes(show)
        agree_frames(String(kind),
            "position" => (ndims(pos) == 3 ? Base.size(pos, 3) : 0),
            "color" => knob_frames(c, n, 4, "$kind.color"),
            "size" => knob_frames(s, n, 1, "$kind.size"),
            "show" => knob_frames(v, n, 1, "$kind.show"))
        sbd = scale_by_distance === nothing ? nothing : NTuple{4,Float64}(scale_by_distance)
        return new(String(kind), pos, c, s, v, mk, label isa Label ? label : texts, sbd)
    end
end

Nodes(kind; position, color = nothing, size = nothing, marker = :disc, label = nothing,
      show = nothing, scale_by_distance = nothing) =
    Nodes(kind, position, color, size, show, marker, label, scale_by_distance)

File: lib/primitives/src/nodes.ts
// Points or sprites: one billboard collection per family, so the whole family is one draw command
// whatever its colours — a billboard's colour, size and visibility are vertex attributes of the
// collection's batch, not separate primitives.
//
// Positions blend every tick; colour, size and visibility switch at the crossing. That split is the
// one assumption the vendored renderer makes about meaning, and it is what makes motion smooth.

import type {
  BillboardCollection, Cartesian3, Color, LabelCollection, NearFarScalar, Scene,
} from "@cesium/engine";
import type { NdArray } from "../../core/src/codec.ts";
import type { WindowInfo } from "../../core/src/windows.ts";
import { at, knob, type Knob, type Slice } from "./knobs.ts";
import { colorOf, WHITE, type CesiumRuntime } from "./paint.ts";
import type { Placement, Timeline } from "../../core/src/windows.ts";
import { markerSprite } from "./sprites.ts";

/** A node family as Julia sends it. */
export interface NodeSpec {
  kind: string;
  /** ECEF metres: `3 x N` static for the window, or `3 x N x count` interpolated across it. */
  position: NdArray;
  color?: unknown;
  size?: unknown;
  show?: unknown;
  /** What to draw each entity with: a stock glyph name, an `assets/<mount>/<file>` path, a
   *  `data:` URI, or the owner-namespaced name of a sprite a peer module registered. */
  marker?: string;
  /** One label per entity, drawn beside the marker. */
  label?: string[];
  /** `[near_m, near_scale, far_m, far_scale]`: markers stay legible close up and shrink far out. */
  scaleByDistance?: number[];
}

interface NodeWindow {
  n: number;
  position: Knob;
  color: Knob | null;
  size: Knob | null;
  show: Knob | null;
}

const DEFAULT_SIZE = 10;

/** One node family: the billboards, their live positions, and what each window said about them. */
export class NodeFamily {
  readonly positions: Cartesian3[] = [];
  private billboards: BillboardCollection | null = null;
  private labels: LabelCollection | null = null;
  private readonly timeline: Timeline<NodeWindow>;
  private n = 0;
  private built = "";
  private scale: NearFarScalar | undefined;
  private readonly scratch: Color;

  readonly kind: string;
  private readonly C: CesiumRuntime;
  private readonly scene: Scene;
  private readonly pickId: (kind: string, idx: number) => object;
  private readonly assetUrl: (path: string) => string | null;

  constructor(
    kind: string,
    C: CesiumRuntime,
    scene: Scene,
    pickId: (kind: string, idx: number) => object,
    assetUrl: (path: string) => string | null,
    timeline: Timeline<NodeWindow>,
  ) {
    this.kind = kind;
    this.C = C;
    this.scene = scene;
    this.pickId = pickId;
    this.assetUrl = assetUrl;
    this.scratch = new C.Color();
    this.timeline = timeline;
  }

  /** Take one window's arrays, rebuilding the collection only when the family's shape changed. */
  onWindow(spec: NodeSpec, win: WindowInfo): void {
    const count = win.count;
    const n = spec.position.shape[spec.position.shape.length - 2] ?? 0;
    const w: NodeWindow = {
      n,
      position: knob(spec.position, { itemLen: 3, n, count, what: `${this.kind}.position` })!,
      color: knob(spec.color, { itemLen: 4, n, count, what: `${this.kind}.color` }),
      size: knob(spec.size, { itemLen: 1, n, count, what: `${this.kind}.size` }),
      show: knob(spec.show, { itemLen: 1, n, count, what: `${this.kind}.show` }),
    };
    // The billboards themselves depend only on the family's shape: how many entities, which glyph,
    // and what each is labelled. Everything else is an attribute write on standing primitives.
    const marker = spec.marker ?? "disc";
    const signature = `${n}|${marker}|${spec.scaleByDistance ?? []}|${
      (spec.label ?? []).join("\u0000")}`;
    if (signature !== this.built) {
      // Sizes from this window's first keyframe, which is what the marker is born at. A billboard
      // built at one size and written to another draws right either way, but an SVG marker is
      // rasterized once, at the size it was born with — see `build`.
      this.build(n, marker, spec.label, spec.scaleByDistance, w.size?.frame(0) ?? null);
      this.built = signature;
    }
    this.n = n;
    this.timeline.install(w, win);
  }

  /**
   * The collection this family draws through. `size` is the first keyframe's sizes, or null where
   * the window sets none.
   *
   * A billboard is born at the size it will be drawn at, rather than at a placeholder the first
   * keyframe corrects. A raster image would not care — it is uploaded at its own resolution and the
   * quad is scaled — but Cesium rasterizes an SVG at the width and height the billboard carries when
   * the image is assigned, and never again. Born at a placeholder, an SVG marker is a small texture
   * stretched over a larger quad for the rest of its life.
   *
   * Two cases still outgrow their raster, both only for SVG: a `size` that grows over the window,
   * and `scaleByDistance` drawing the marker larger than its base size close up. Give such a family
   * a raster image instead.
   */
  private build(n: number, marker: string, labels?: string[], scaleByDistance?: number[],
                size: Slice | null = null): void {
    const { C, scene } = this;
    this.destroyPrimitives();
    this.positions.length = 0;
    this.scale = scaleByDistance && scaleByDistance.length === 4
      ? new C.NearFarScalar(scaleByDistance[0], scaleByDistance[1], scaleByDistance[2], scaleByDistance[3])
      : undefined;
    const billboards = scene.primitives.add(new C.BillboardCollection({ scene })) as BillboardCollection;
    const image = markerSprite(marker, this.assetUrl);
    for (let i = 0; i < n; i++) {
      const position = new C.Cartesian3();
      this.positions.push(position);
      const px = size ? at(size, i) : DEFAULT_SIZE;
      billboards.add({
        position,
        image,
        // A stock glyph is white, so the billboard's own colour is what tints it and per-entity
        // colour needs no second image. That tint multiplies a supplied image too, which the default
        // white leaves as it was drawn.
        //
        // `size` writes both width and height, so an image is drawn square whatever its own shape.
        width: px, height: px,
        scaleByDistance: this.scale,
        id: this.pickId(this.kind, i),
      });
    }
    this.billboards = billboards;
    if (labels && labels.length) {
      const collection = scene.primitives.add(new C.LabelCollection({ scene })) as LabelCollection;
      for (let i = 0; i < n; i++) {
        collection.add({
          position: this.positions[i],
          text: labels[i] ?? "",
          font: "13px sans-serif",
          fillColor: C.Color.WHITE,
          outlineColor: C.Color.BLACK,
          outlineWidth: 2,
          style: C.LabelStyle.FILL_AND_OUTLINE,
          pixelOffset: new C.Cartesian2(0, -14),
          verticalOrigin: C.VerticalOrigin.BOTTOM,
        });
      }
      this.labels = collection;
    }
  }

  /** True when this family's positions interpolate rather than stand still for the window. */
  get moving(): boolean {
    return this.timeline.latest?.position.keyframed ?? false;
  }

  /** The pick stamp of entity `idx`, for a module drawing something anchored to it. */
  pickIdAt(idx: number): object | undefined {
    if (idx < 0 || idx >= this.n) return undefined;
    return this.billboards?.get(idx)?.id as object | undefined;
  }

  /**
   * Whether entity `idx` is drawn. Read off the billboard rather than kept beside it, so what an
   * anchored primitive hides with is the mask this family is actually drawing. A billboard is born
   * shown and takes its first mask at the first keyframe.
   */
  shownAt(idx: number): boolean | undefined {
    if (idx < 0 || idx >= this.n) return undefined;
    return this.billboards?.get(idx)?.show ?? true;
  }

  /** Colour, size and visibility at the keyframe `p` names — attribute writes, never a rebuild. */
  onKeyframe(where: Placement | null): void {
    const place = this.timeline.at(where);
    const bb = this.billboards;
    if (!place || !bb) return;
    const { w, k } = place;
    const color = w.color?.frame(k) ?? null;
    const size = w.size?.frame(k) ?? null;
    const show = w.show?.frame(k) ?? null;
    for (let i = 0; i < this.n; i++) {
      const b = bb.get(i);
      b.color = colorOf(this.C, color, i, WHITE, this.scratch);
      const px = size ? at(size, i) : DEFAULT_SIZE;
      b.width = px;
      b.height = px;
      b.show = show ? at(show, i) !== 0 : true;
      if (this.labels) this.labels.get(i).show = b.show;
    }
  }

  /**
   * Blend positions from the keyframe `atA` names toward the one `atB` names — the one knob that
   * interpolates rather than switches. `atB` falls back to `atA` at the run's last frame and for a
   * static family, which have nothing to blend toward.
   */
  onFrame(atA: Placement | null, atB: Placement | null, alpha: number): void {
    const a = this.timeline.at(atA);
    if (!a || !this.billboards) return;
    const b = this.timeline.at(atB) ?? a;
    const pa = a.w.position.frame(a.k);
    const pb = b.w.position.frame(b.k) ?? pa;
    if (!pa || !pb) return;
    const t = 1 - alpha;
    for (let i = 0; i < this.n; i++) {
      const p = this.positions[i];
      p.x = at(pa, i, 0) * t + at(pb, i, 0) * alpha;
      p.y = at(pa, i, 1) * t + at(pb, i, 1) * alpha;
      p.z = at(pa, i, 2) * t + at(pb, i, 2) * alpha;
      // The setter clones into the billboard's own storage and only marks dirty on a real move, so a
      // static family costs a comparison per entity and no upload.
      this.billboards.get(i).position = p;
      if (this.labels) this.labels.get(i).position = p;
    }
  }

  private destroyPrimitives(): void {
    for (const prim of [this.billboards, this.labels]) {
      if (prim) {
        try { this.scene.primitives.remove(prim); } catch { /* already gone */ }
      }
    }
    this.billboards = null;
    this.labels = null;
  }

  destroy(): void {
    this.destroyPrimitives();
    this.timeline.clear();
    this.built = "";
  }
}
